src/main/java/duke/storage/Storage.java

```diff
@@ -0,0 +1,138 @@
+                    break;
+                case "E":
+                    task = new Event(description, time);
+                    break;
+                case "D":
+                    task = new Deadline(description, time);
+                    break;
+                default:
+                    throw new JsonParseException("Invalid task type.");
+                }
+            } catch (DukeException e) {
+                throw new JsonParseException("Invalid date format.");
+            }
+            if (elemObject.get("isDone").getAsBoolean()) {
+                task.markAsDone();
+            }
+            tasks.add(task);
+        }
+        return new TaskList(tasks);
+    };
+
+    static {
+        GSON_BUILDER.registerTypeAdapter(TaskList.class, TASK_LIST_SERIALIZER);
+        GSON_BUILDER.registerTypeAdapter(TaskList.class, TASK_LIST_DESERIALIZER);
+    }
+
+    private final String filePath;
+    private final Gson gsonObject;
+
+    /**
+     * Constructor for Storage.
+     * @param filePath Relative String filepath to the input json file.
+     */
+    public Storage(String filePath) {
+        this.filePath = filePath;
+        this.gsonObject = GSON_BUILDER.create();
+    }
+
+    private static String parseDate(JsonElement elem) throws JsonParseException {
+        if (elem == null) {
+            return null;
+        }
+        JsonObject obj = elem.getAsJsonObject();
+        try {
+            int year = obj.get("year").getAsInt();
+            int month = obj.get("month").getAsInt();
+            int date = obj.get("day").getAsInt();
+            return String.format("%04d-%02d-%02d", year, month, date);
+        } catch (NullPointerException e) {
+            throw new JsonParseException("Invalid date format.");
+        }
+    }
+
+    /**
+     * Saves TaskList in Json format to the filepath supplied to the constructor.
+     * @param taskList TaskList to save.
+     * @throws DukeException If there is any IOException.
+     */
+    public void save(TaskList taskList) throws DukeException {
+        try (FileWriter fw = new FileWriter(this.filePath)) {
+            this.gsonObject.toJson(taskList, TaskList.class, fw);
+        } catch (IOException e) {
+            throw DukeException.Errors.FILE_WRITE_ERROR.create();
+        }
+    }
+
+    /**
+     * Loads TaskList from a previously saved Json file. Uses filepath supplied to constructor.
+     * @return TaskList parsed from Json file.
+     * @throws DukeException If any parsing errors occur (wrong format, etc).
+     */
+    public TaskList load() throws DukeException {
+        Path path = Paths.get(filePath);
+        if (!path.toFile().isFile()) {
+            return new TaskList();
+        }
+        try (Reader reader = Files.newBufferedReader(path)) {
+            return gsonObject.fromJson(reader, TaskList.class);
+        } catch (IOException e) {
+            throw DukeException.Errors.FILE_READ_ERROR.create();
+        }
+    }
+

```

Comment:
Impressive implementation! Although there are many lines of code, it is still very readable.